File: java/src/tp_1/GestorEstudiantes.java
package tp_1;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

public class GestorEstudiantes {
	private Set<Estudiante> estudiantes;
	
	public GestorEstudiantes() {
		this.estudiantes = new TreeSet<>();
	}
	
	public GestorEstudiantes(Estudiante estudiante) {
		this();
		this.estudiantes.add(estudiante);
	}
	
	public GestorEstudiantes(Estudiante[] estudiantes) {
		this();
		for (int i = 0; i < estudiantes.length; i++) {
			this.estudiantes.add(estudiantes[i]);			
		}
	}
	
	public GestorEstudiantes(Set<Estudiante> estudiantes) {
		this();
		this.estudiantes.addAll(estudiantes);
	}
	
	public GestorEstudiantes(List<Estudiante> estudiantes) {
		this();
		this.estudiantes.addAll(estudiantes);
	}
	
	public Estudiante buscarEstudiantePorDNI(int dni) {
		for (Estudiante estudiante : this.estudiantes) {
			if(estudiante.getDni() == dni) {
				return estudiante;
			}
		}
		return null;
	}
	
	public List<Estudiante> listarEstudiantesPromedioAprobado(){
		List<Estudiante> estudiantesRet = new ArrayList<>();
		
		for (Estudiante estudiante : this.estudiantes) {
			if(estudiante.calcularPromedio() >= 7) {
				estudiantesRet.add(estudiante);
			}
		}
		return estudiantesRet;
	}
}

File: go/gestorEstudiantesEnGo/Estudiante/gestorEstudiantes.go
package Estudiante

// GestorEstudiantes usa un map con clave DNI
type GestorEstudiantes struct {
	estudiantes map[int]*Estudiante
}

// Constructor vacío
func NuevoGestorEstudiantes() *GestorEstudiantes {
	return &GestorEstudiantes{
		estudiantes: make(map[int]*Estudiante),
	}
}

// En go no existe la sobrecarga, asi que debemos crear dos métodos distintos
func NuevoGestorConEstudiante(e *Estudiante) *GestorEstudiantes {
	g := NuevoGestorEstudiantes()
	g.estudiantes[e.DNI] = e
	return g
}

// Constructor con slice de estudiantes
func NuevoGestorConSlice(slice []*Estudiante) *GestorEstudiantes {
	g := NuevoGestorEstudiantes()
	for _, e := range slice {
		g.estudiantes[e.DNI] = e
	}
	return g
}

// Buscar estudiante por DNI
func (g *GestorEstudiantes) BuscarEstudiantePorDNI(dni int) *Estudiante {
	if est, existe := g.estudiantes[dni]; existe {
		return est
	}
	return nil
}

// Listar estudiantes con promedio >= 7
func (g *GestorEstudiantes) ListarEstudiantesPromedioAprobado() []*Estudiante {
	var aprobados []*Estudiante
	for _, e := range g.estudiantes {
		if e.CalcularPromedio() >= 7 {
			aprobados = append(aprobados, e)
		}
	}
	return aprobados
}
